src/input/sources/hls.rs

```diff
@@ -0,0 +1,139 @@
+use std::{
+    io::{ErrorKind as IoErrorKind, Result as IoResult, SeekFrom},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::Bytes;
+use futures::StreamExt;
+use pin_project::pin_project;
+use reqwest::{header::HeaderMap, Client};
+use serenity::async_trait;
+use stream_lib::{DownloadStream, Event};
+use symphonia_core::io::MediaSource;
+use tokio::io::{AsyncRead, AsyncSeek, ReadBuf};
+use tokio_util::io::StreamReader;
+
+use crate::input::{
+    AsyncAdapterStream,
+    AsyncMediaSource,
+    AudioStream,
+    AudioStreamError,
+    Compose,
+    Input,
+};
+
+/// Lazy HLS stream
+#[derive(Debug)]
+pub struct HlsRequest {
+    /// HLS downloader
+    pub hls: Option<DownloadStream>,
+}
```

Comment:
I think this would be closer to the intent of `Compose` if `HlsRequest` just held onto a `Client` and `String` like HttpRequests do -- this is generally fine since reqwest clients are internally `Arc`'d. `Compose` structs are meant as a 'blueprint' of sorts to create usable streams. That way you create a `DownloadStream`/`HlsStream` every time you call `HlsRequest::create_async`.

This also prevents the 'can only use once' issue below.